torch_geometric/nn/conv/hgt_conv.py

```diff
@@ -143,40 +177,105 @@ def forward(
         """
         H, D = self.heads, self.out_channels // self.heads
 
-        k_dict, q_dict, v_dict, out_dict = {}, {}, {}, {}
+        k_dict, q_dict, v_dict = {}, {}, {}
+        out_dict = defaultdict(list)
+
+        # parallelize over node-types
+        # compute K, Q, V over node-types
+        k_dict = {
+            node_type: k_j.view(-1, H, D)
+            for node_type, k_j in self.k_lin(x_dict).items()
+        }
+        q_dict = {
+            node_type: q_j.view(-1, H, D)
+            for node_type, q_j in self.q_lin(x_dict).items()
+        }
+        v_dict = {
+            node_type: v_j.view(-1, H, D)
+            for node_type, v_j in self.v_lin(x_dict).items()
+        }
+
+        # parallelize over edge-types
+        a_rels = [
+            self.a_rel['__'.join(edge_type)] for edge_type in self.edge_types
+        ]
+        m_rels = [
+            self.m_rel['__'.join(edge_type)] for edge_type in self.edge_types
+        ]
+
+        k_ins = []
+        v_ins = []
+        count = 0
+        trans_ptr = [count]
+        for src_type in self.src_types:
+            k_src = k_dict[src_type]
+            v_src = v_dict[src_type]
+            k_ins.append(k_src.view(-1, D))
+            v_ins.append(v_src.view(-1, D))
+            for i in range(H):
+                count += k_src.size(0)
+                trans_ptr.append(count)
+        trans_ptr = torch.tensor(trans_ptr)
+        print("Trans", trans_ptr.shape, trans_ptr)
+        a_rel, m_rel = torch.cat(a_rels), torch.cat(m_rels)
+        k_out = pyg_lib.ops.segment_matmul(torch.cat(k_ins), trans_ptr,
+                                           a_rel).view(-1, H, D)
+
+        v_out = pyg_lib.ops.segment_matmul(torch.cat(v_ins), trans_ptr,
+                                           m_rel).view(-1, H, D)
+        print("K_out", k_out.shape)
+        node_slices = make_node_slices({
+            n_type: k_out[trans_ptr[i]:trans_ptr[i + 1]].shape[0]
+            for i, n_type in enumerate(self.node_types)
+        })
+        print("node_slices", node_slices)
+        # combine edge_index dict into single tensor
+        q_list = []
+        p_rels = []
+        edge_index_list = []
+        for e_type in self.edge_types:
+            indices = edge_index_dict[e_type]
+            # (TODO) Add support for SparseTensor w/o converting?
+            convert = isinstance(indices, SparseTensor)
+            if convert:
+                # convert to COO
+                dst, src, _ = indices.coo()
+                indices = torch.cat((src.view(1, -1), dst.view(1, -1)))
+            edge_index_list.append(
+                offset_edge_idx(node_slices, e_type, indices))
+            q_list.append(q_dict[e_type[-1]])
+            p_rels.append(self.p_rel['__'.join(e_type)].view(-1, 1))
+
+        q = torch.cat(q_list)
+        print('Q', q.shape)
+        p = group(p_rels, self.group).view(-1)
```

Comment:
```suggestion
        # Concatenate all dst node tensors.
        q_list = []
        dst_offset = {}
        offset = 0
        for node_type in self.node_types:
            q_list.append(q_dict[node_type])
            dst_offset[node_type] = offset
            offset += q_dict[node_type].size(0)
        q = torch.cat(q_list)

        # list of edges and edge tensors.
        edge_index_list = []
        edge_src_offset = 0
        p_rels = []


        # list of src node tensors for each edge
        k_ins = []
        v_ins = []
        count = 0
        trans_ptr = [count]


        for edge_type in self.edge_types:
            src_type, _, dst_type = edge_type

            # Collect src node tesnors.
            k_src = k_dict[src_type]
            v_src = v_dict[src_type]
            k_ins.append(k_src.view(-1, D))
            v_ins.append(v_src.view(-1, D))
            for _ in range(H):
                count += k_src.size(0)
                trans_ptr.append(count)

            # Collect edges and edge tensors.
            p_rels.append(self.p_rel['__'.join(edge_type)].view(-1, 1))
            indices = edge_index_dict[edge_type]
            # (TODO) Add support for SparseTensor w/o converting?
            convert = isinstance(indices, SparseTensor)
            if convert:
                # convert to COO
                dst, src, _ = indices.coo()
                indices = torch.cat((src.view(1, -1), dst.view(1, -1)))
            # Add offset to edge indices. Offset needs to be added as
            # the src and dst node tensors are concatenated.
            edge_index_list.append(offset_edge(indices, (edge_src_offset, dst_offset[dst_type])))

            # Update offsets.
            edge_src_offset += k_src.size(0)

        # Concatenate all src node tensors.
        trans_ptr = torch.tensor(trans_ptr)
        a_rel, m_rel = torch.cat(a_rels), torch.cat(m_rels)
        k_out = pyg_lib.ops.segment_matmul(torch.cat(k_ins), trans_ptr,
                                           a_rel).view(-1, H, D)
        v_out = pyg_lib.ops.segment_matmul(torch.cat(v_ins), trans_ptr,
                                           m_rel).view(-1, H, D)

        p = group(p_rels, self.group).view(-1)
```
Here's the suggested changes as we discussed. 
I am still not sure about how we compute `p`. But i'll leave that for later.

Reply:
thanks! it makes alot more sense to me now seeing the code :)